client/state/analytics/actions/enhance-with-site-main-product.js

```diff
@@ -0,0 +1,46 @@
+import { hasGSuiteWithUs } from 'calypso/lib/gsuite';
+import { hasTitanMailWithUs } from 'calypso/lib/titan';
+import { ANALYTICS_PAGE_VIEW_RECORD } from 'calypso/state/action-types';
+import isDomainOnly from 'calypso/state/selectors/is-domain-only-site';
+import { getDomainsBySiteId } from 'calypso/state/sites/domains/selectors';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+
+/**
+ * Enhances any Redux action that denotes the recording of a page view analytics event with an additional property
+ * to specifies the main product of this site:
+ * `domain` - domain only, without email subscription
+ * `email` - domain only with email subscription
+ * `site` - regular site
+ *
+ * @param {object} action - Redux action as a plain object
+ * @param {Function} getState - Redux function that can be used to retrieve the current state tree
+ * @returns {import('redux').AnyAction} the new Redux action
+ * @see client/state/utils/withEnhancers
+ */
+export function enhanceWithSiteMainProduct( action, getState ) {
+	if ( action.type !== ANALYTICS_PAGE_VIEW_RECORD ) {
+		return action;
+	}
+
+	const state = getState();
+	const siteId = getSelectedSiteId( state );
+	let mainProduct = 'site';
+
+	if ( isDomainOnly( state, siteId ) ) {
+		mainProduct = 'domain';
+
+		// Domain only site should only have one single domain non-wpcom domain.
+		const nonWPCOMDomain = getDomainsBySiteId( state, siteId ).find(
+			( domain ) => ! domain.isWPCOMDomain
+		);
+
+		if ( hasTitanMailWithUs( nonWPCOMDomain ) || hasGSuiteWithUs( nonWPCOMDomain ) ) {
+			mainProduct = 'email';
+		}
+	}
+
+	const updatedAction = action;
+	updatedAction.meta.analytics[ 0 ].payload.site_main_product = mainProduct;
```

Comment:
Removed use of Lodash `set()` and returning a new copy, as recommended.

Reply:
Thanks, but this will still mutate the original object.

Try this out in your browser console:

```
const a = { something: 1 };
const b = a;
b.something++;
console.log(a.something)
```

and you'll see `b.something` also increased to 2. The reason is because in JavaScript objects are assigned by reference, and not by value, so `b` will still point to `a`, and if we change one of them, the other will be changed as well. Understanding how we [compare objects](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Working_with_Objects#comparing_objects) can also help understand that.

In order to fix that, we should actually create a copy of the object. There's a recent native function for that which we still haven't leveraged in Calypso - [`structuredClone()`](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone) - that might have a bit of limitations, for example, it's available only in most recent Node.js versions (17+ I believe). It should be supported by modern browsers, though. Alternatively, we could just declare a new object and spread the old one into it, but note that this needs to be done for all properties in depth, as the same reference problem can occur for another inner object that's part of the primary object.

Reply:
Another classic solution to cloning is to do:
```js
const updatedAction = JSON.parse( JSON.stringify( action ) );
```

Looking at our browserlist configs [here](https://github.com/Automattic/wp-calypso/blob/trunk/package.json#L24-L54) and [here](https://github.com/WordPress/gutenberg/blob/9d1d8e3e6ef956a5f2aa9d5863f535c7f55862a6/packages/browserslist-config/index.js) (which should result in something like [this](https://browsersl.ist/?q=%3E+1%25%2C+last+1+Android+versions%2C+last+1+ChromeAndroid+versions%2C+last+2+Chrome+versions%2C+last+2+Firefox+versions%2C+last+2+Safari+versions%2C+last+3+iOS+versions%2C+last+2+Edge+versions%2C+last+2+Opera+versions)), `structuredClone` seems to be blocked by Safari 14.

Don't know how concerned we are with that, but there is no obvious benefit to what `structuredClone` supports [in terms of data types](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) over the `JSON` approach. So for our purposes, I think the latter _should_ be fine (and it's probably more convenient than spreading with an object structure as deep as this).

Reply:
Agreed, happy to accept this workaround for an action since it will always be a simple object.